lib/config/decrypt.spec.ts

```diff
@@ -208,5 +216,92 @@ describe('config/decrypt', () => {
         CONFIG_VALIDATION,
       );
     });
+
+    it('handles EC/AES-GCM encryption', async () => {
+      GlobalConfig.set({ privateKey: toBase64(privateEc) });
+      config.encrypted = {
+        token:
+          'eyJrIjp7Imt0eSI6IkVDIiwieCI6IlM0TFhnVUJHMUNlSXVUU1ZiZEg3MHZpSjV6dUxaRFNydFJ2ZUhVTHBBYnBMTXAwQ0tyeVozWm5RN1lKQjh3N2MiLCJ5IjoiZVpQNFZzT1hXZGxRYkNWYWxSakVWNEFzTFBxYmpKdW1mR1dfTzlZbG9HM2dqbzd5enJfUXQtbXl2SG5LdkZheiIsImNydiI6IlAtMzg0In0sImkiOiJNRDVMb2hUMmh2VS9SYXA3IiwibSI6IkptMERpUmNtbEFEU2JjeVdnWmJlUXl4QkdUbU55aVlkV2NDa0tyVnhQZStTdEdMMlBwWT0ifQ==',
+      };
+
+      const res = await decryptConfig(config, 'some/def');
+      expect(res.encrypted).toBeUndefined();
+      expect(res.token).toBe('123');
+      await expect(decryptConfig(config, 'abc/defg')).rejects.toThrow(
+        CONFIG_VALIDATION,
+      );
+    });
+  });
+
+  describe('tryDecryptEc', () => {
+    it('decrypts', async () => {
+      // Generate a new key pair in the browser for each encryption,
+      // this private key will be thrown away after encryption,
+      // so that only Renovate can decrypt with the private key.
+      const browserKeyPair = await crypto.subtle.generateKey(
+        { name: 'ECDH', namedCurve: 'P-384' },
+        false,
+        ['deriveKey'],
+      );
+
+      // `d` is the private key, `x` and `y` are the public key
+      const { d, ...renovatePublicKey } = { ...renovatePrivateKey };
+
+      // import Renovate's public key
+      const importedRenovatePublicKey = await crypto.subtle.importKey(
+        'jwk',
+        renovatePublicKey,
+        { name: 'ECDH', namedCurve: 'P-384' },
+        false,
+        [],
+      );
+
+      // derive shared secret from our private key and Renovate's public key
+      const pw = await crypto.subtle.deriveKey(
```

Comment:
This looks **cryptographically problematic** to me.

The shared secret derived by an ECDH key agreement is not uniformly random. `deriveKey` doesn't perform any key stretching or hashing on the shared secret. It is basically the same as calling `deriveBits` to get a byte array and then passing it to `importKey` to use as a symmetric key with AES. P-384 generates more output bits than needed, some part will simply be dropped to fit the 256 bits needed for AES.

To derive a cryptographically strong key for AES-GCM, the shared secret needs to be processed via [HKDF](https://en.wikipedia.org/wiki/HKDF). An example how to do this can be found here: https://stackoverflow.com/a/67942717/589259

Also see:
- https://github.com/w3c/webcrypto/issues/193
- https://security.stackexchange.com/questions/201284/ecdh-and-static-key-encryption
- https://crypto.stackexchange.com/questions/30367/ecdh-security-when-no-kdf-is-used

Reply:
🤔 didn't know that. do we need a separate value for salt, or can we reuse the aes-gcm iv?
We should try to make the encypted object as small as possible but of cause as secure as possible.

Reply:
Salt and IV are similar in some ways but also need to fulfill different requirements depending on what you want to achieve. I'd suggest to not simply reuse the same values. What you could do **in this particular case**:

Call `deriveBits({ name: "HKDF", ...` (as shown in the stackoverflow example above) and produce bits for both the key _and_ the VI. You can use the `iv` (as it is currently named in your code) as a salt value for the HKDF. When calling `deriveBits`, increase the `length` param by the 96 bits you need for the IV, so 256 + 96 bits overall, then use the trailing 96 bits for the IV.

This way you don't need to prepend it to the ciphertext and the HKDF ensures it's uniformly random for use with a symmetric cipher like AES. If you change the salt, the encryption key and IV will change as well. This is generally desirable because it ensures that if you encrypt multiple texts (with different salt values in the HKDF call), a unique encryption key and IV will be used as well.